src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -270,21 +280,75 @@ private void registerNestedSubclassToOuterSuperClasses(String qualifiedClassName
+                                                         String outerTypeDeclName) {
+        final String superClassName = CheckUtil.getShortNameOfAnonInnerClass(literalNewAst);
+        final ClassDesc superClassDesc = getNearestClassWithSameName(
+            superClassName, outerTypeDeclName, true);
+        if (superClassDesc == null) {
+            Optional.ofNullable(innerClasses.get(superClassName))
+                .ifPresent(ClassDesc::registerSuperClassOfAnonymousInnerClass);
+        }
+        else {
+            superClassDesc.registerSuperClassOfAnonymousInnerClass();
+        }
+    }
+
     /**
      * Checks if there is a class with same name.
      *
+     * <p>The parameter {@code shouldSearchAnonSuperClass} exists because if the class
+     * being searched for is the super class of anonymous inner class, the rules of evaluation are
+     * a bit different, consider the following example-
+     * <pre>
+     * {@code
+     * public class Main {
+     *     static class a {
+     *         static class b {
+     *         }
+     *     }
+     *
+     *     class j {
+     *         a.b object = new a.b() { // Object of Main.j.a.b and not of Main.a.b
+     *         };
+     *
+     *         static class a {
+     *             static class b {
+     *             }
+     *         }
+     *     }
+     * }
+     * }
+     * </pre>
+     * If the parameter wasn't set to {@code true} then the logic would have falsely evaluated
+     * {@code Main.a.b} to be the super class of the anonymous inner class. Set this to
+     * {@code false} only if the class being searched for is not the super class of anonymous
+     * inner class.
+     *
      * @param className name of the class
-     * @param superClassName name of the super class
+     * @param superOrOuterClassName name of the super class or the outer class
+     * @param shouldSearchAnonSuperClass should search for the super class of anonymous inner class
      * @return true if there is another class with same name.
      * @noinspection CallToStringConcatCanBeReplacedByOperator
      */
-    private ClassDesc getNearestClassWithSameName(String className, String superClassName) {
+    private ClassDesc getNearestClassWithSameName(String className, String superOrOuterClassName,
+                                                  boolean shouldSearchAnonSuperClass) {
         final String dotAndClassName = PACKAGE_SEPARATOR.concat(className);
         return innerClasses.entrySet().stream()
                 .filter(entry -> entry.getKey().endsWith(dotAndClassName))
                 .map(Map.Entry::getValue)
                 .min((first, second) -> {
-                    return getClassDeclarationNameMatchingCountDiff(superClassName, first, second);
+                    return getClassDeclarationNameMatchingCountDiff(superOrOuterClassName, first,
+                         second, shouldSearchAnonSuperClass);
                 })
                 .orElse(null);
```

Comment:
why you return null here if in registerAnonymousInnerClassToSuperClass method you check for null anyway?

Reply:
Done, used Optional